docs(pt-br): tradução da referencia useState (#782)
reactjs/pt-br.react.dev: src/content/reference/react/useState.md

</Sandpack>

> Note que se você chamar uma função `set` durante a renderização, ela deve estar dentro de uma condição como `prevCount !== count`, e deve haver uma chamada como `setPrevCount(count)` dentro da condição. Caso contrário, seu componente renderizaria em um loop até travar. Além disso, você só pode atualizar o state do componente *atualmente renderizando* desta forma. Chamar a função `set` de *outro* componente durante a renderização é um erro. Finalmente, sua chamada `set` ainda deve [atualizar o state sem mutação](#updating-objects-and-arrays-in-state) -- isso não significa que você pode quebrar outras regras das [funções puras.](/learn/keeping-components-pure)

Reviewer: ```suggestion
Note que se você chamar uma função `set` durante a renderização, ela deve estar dentro de uma condição como `prevCount !== count`, e deve haver uma chamada como `setPrevCount(count)` dentro da condição. Caso contrário, seu componente renderizaria em um loop até travar. Além disso, você só pode atualizar o state do componente que está *renderizando* desta forma. Chamar a função `set` de *outro* componente durante a renderização é um erro. Por fim, sua chamada `set` deve [atualizar o state sem mutação](#updating-objects-and-arrays-in-state) -- isso não significa que você pode quebrar outras regras das [funções puras.](/learn/keeping-components-pure)
```

> Este padrão pode ser difícil de entender e geralmente é melhor evitá-lo. No entanto, é melhor do que atualizar o state em um efeito. Quando você chama a função `set` durante a renderização, o React renderizará esse componente imediatamente após seu componente sair com uma instrução `return`, e antes de renderizar os filhos. Desta forma, os filhos não precisam renderizar duas vezes. O resto da função do seu componente ainda será executado (e o resultado será descartado). Se sua condição estiver abaixo de todas as chamadas de Hook, você pode adicionar um `return;` antecipado para reiniciar a renderização mais cedo.

Reviewer: ```suggestion
Este padrão pode ser difícil de entender e geralmente é melhor evitá-lo. No entanto, é melhor do que atualizar o state em um effect. Quando você chama a função `set` durante a renderização, o React irá renderizar esse componente imediatamente após seu componente sair com uma instrução `return`, e antes de renderizar os filhos. Desta forma, os filhos não precisam renderizar duas vezes. O resto da função do seu componente ainda será executado (e o resultado será descartado). Se sua condição estiver abaixo de todas as chamadas de Hook, você pode adicionar um `return;` antecipado para reiniciar a renderização mais cedo.
```

---

### Atualizei o state, mas a tela não atualiza {/*ive-updated-the-state-but-the-screen-doesnt-update*/}

> O React **ignorará sua atualização se o próximo state for igual ao state anterior**, determinado por uma comparação [`Object.is`](https://developer.mozilla.org/pt-BR/docs/Web/JavaScript/Reference/Global_Objects/Object/is). Isso geralmente acontece quando você altera um objeto ou um array no state diretamente:

Reviewer: ```suggestion
O React **ignorará sua atualização se o próximo state for igual ao state anterior**, determinado por uma comparação [`Object.is`](https://developer.mozilla.org/pt-BR/docs/Web/JavaScript/Reference/Global_Objects/Object/is). Isso geralmente acontece quando você altera um objeto ou um array no state diretamente:
```
```suggestion
O React **irá ignorar sua atualização se o próximo state for igual ao state anterior**, determinado por uma comparação [`Object.is`](https://developer.mozilla.org/pt-BR/docs/Web/JavaScript/Reference/Global_Objects/Object/is). Isso geralmente acontece quando você altera um objeto ou um array no state diretamente:
```

```js

> obj.x = 10;  // 🚩 Errado: mutando um objeto existente

Reviewer: ```suggestion
obj.x = 10;  // 🚩 Errado: fez mutação em um objeto existente
```

setObj(obj); // 🚩 Não faz nada
```

> Você mutou um objeto `obj` existente e o passou de volta para `setObj`, então o React ignorou a atualização. Para corrigir isso, você precisa garantir que sempre está _substituindo_ objetos e arrays no state em vez de _mutá-los_:

Reviewer: ```suggestion
Você alterou um objeto `obj` existente e o passou de volta para `setObj`, então o React ignorou a atualização. Para corrigir isso, você precisa garantir que sempre está _substituindo_ objetos e arrays no state em vez de _realizar uma mutação_:
```

---

### Minha função inicializadora ou atualizadora é executada duas vezes {/*my-initializer-or-updater-function-runs-twice*/}

No [Modo Estrito](/reference/react/StrictMode), o React chamará algumas de suas funções duas vezes em vez de uma:

```js {2,5-6,11-12}
function TodoList() {
  // Esta função de componente será executada duas vezes para cada renderização.

  const [todos, setTodos] = useState(() => {
    // Esta função inicializadora será executada duas vezes durante a inicialização.
    return createTodos();
  });

  function handleClick() {
    setTodos(prevTodos => {
      // Esta função atualizadora será executada duas vezes para cada clique.
      return [...prevTodos, createTodo()];
    });
  }
  // ...
```

Isso é esperado e não deve quebrar seu código.

Este comportamento **apenas de desenvolvimento** ajuda você a [manter os componentes puros.](/learn/keeping-components-pure) O React usa o resultado de uma das chamadas e ignora o resultado da outra chamada. Desde que suas funções de componente, inicializadoras e atualizadoras sejam puras, isso não deve afetar sua lógica. No entanto, se elas forem acidentalmente impuras, isso ajuda você a perceber os erros.

> Por exemplo, esta função atualizadora impura muta um array no state:

Reviewer: ```suggestion
Por exemplo, esta função atualizadora impura realiza mutação em um array no state:
```

```js {2,3}
setTodos(prevTodos => {

>   // 🚩 Erro: mutando o state

Reviewer: ```suggestion
  // 🚩 Erro: realiza mutação no state
```

  prevTodos.push(createTodo());
});
```

Como o React chama sua função atualizadora duas vezes, você verá que o todo foi adicionado duas vezes, então você saberá que há um erro. Neste exemplo, você pode corrigir o erro [substituindo o array em vez de mutá-lo](#updating-objects-and-arrays-in-state):

```js {2,3}
setTodos(prevTodos => {
  // ✅ Correto: substituindo por um novo state
  return [...prevTodos, createTodo()];
});
```

Agora que esta função atualizadora é pura, chamá-la uma vez extra não faz diferença no comportamento. É por isso que o React chamar duas vezes ajuda você a encontrar erros. **Apenas funções de componente, inicializadoras e atualizadoras precisam ser puras.** Manipuladores de eventos não precisam ser puros, então o React nunca chamará seus manipuladores de eventos duas vezes.

Leia [mantendo componentes puros](/learn/keeping-components-pure) para saber mais.

---

> ### Estou tentando definir o state para uma função, mas ela é chamada em vez disso {/*im-trying-to-set-state-to-a-function-but-it-gets-called-instead*/}

Reviewer: ```suggestion
### Estou tentando definir o state para uma função, mas em vez disso ela é executada {/*im-trying-to-set-state-to-a-function-but-it-gets-called-instead*/}
```

}
```

> Porque você está passando uma função, o React assume que `someFunction` é uma [função inicializadora](#avoiding-recreating-the-initial-state), e que `someOtherFunction` é uma [função atualizadora](#updating-state-based-on-the-previous-state), então tenta chamá-las e armazenar o resultado. Para realmente *armazenar* uma função, você precisa colocar `() =>` antes delas em ambos os casos. Então o React armazenará as funções que você passa.

Reviewer: ```suggestion
Porque você está passando uma função, o React assume que `someFunction` é uma [função inicializadora](#avoiding-recreating-the-initial-state), e que `someOtherFunction` é uma [função atualizadora](#updating-state-based-on-the-previous-state), então tenta chamá-las e armazena o resultado. Para realmente *armazenar* uma função, você precisa colocar `() =>` antes delas em ambos os casos. Então o React irá armazenar as funções que você passa.
```
